### ISSUE TITLE
(GLES3) Zelda: Ocarina of Time - Transparent Link

### ISSUE THREAD

**Issue by orbea** [loganmc10/GLupeN64#74]
OS: `Slackware64-current`
`mesa-2016.10.07_185be15_master-x86_64-1_git`
`RetroArch-2016.10.09_9561b55_master-x86_64-1_git`
`GLupeN64-2016.10.09_fac6464_master-x86_64-1_git`

In certain areas such as the path in Kokiro Village Link will become partially transparent when using GLES3. This does not occur with GLES2 and I have tried both old commits and older mesa builds which also reproduced it meaning its been around for a while.

How do I build upstream with GLES3? I only see how for GLES2.

Apitrace - http://ks392457.kimsufi.com/orbea/stuff/trace/libretro/Zelda%20OoT%2064_GLES3.trace.xz

![1](http://ks392457.kimsufi.com/orbea/stuff/pics/scrots/libretro/glupen64/Zelda%20OoT%2064-161009-182648.png)

**Comment by orbea** [comment]
Disabling Frame Buffer Emulation works around it.

**Comment by orbea** [comment]
This does not occur for me with Android despite the other issue.

**Comment by loganmc10** [comment]
Are you able to provide a save state that puts Link in a position where this happens? I don't have s controller I can plug into my laptop handy to test this

**Comment by orbea** [comment]
Here.

http://ks392457.kimsufi.com/orbea/stuff/games/retroarch/sstates/GLupeN64/Zelda%20OoT%2064.state1.xz

**Comment by loganmc10** [comment]
![screenshot at 2016-10-10 12 54 21](https://cloud.githubusercontent.com/assets/848146/19247548/f216dbfc-8ee8-11e6-8ce3-6c3b4775a217.png)

This is what that save state looks like for me with GLES3 on my laptop (Nvidia drivers, I don't have mesa installed right now unfortunately. I will try tomorrow on my laptop that runs Linux inside VMware, since that uses Mesa).

As far as testing this with upstream, I'm not sure, for GLideN64 it should be as simple as modifying https://github.com/gonetz/GLideN64/blob/master/src/CMakeLists.txt and replacing references to "GLES2" with "GLES3". That's probably all you would need to do.

**Comment by orbea** [comment]
Can you reproduce it by replaying the apitrace? I can't on my ntel laptop which I think should work with GLES3. I will have to try the actual game on it later.

**Comment by orbea** [comment]
I can't reproduce it on intel.

**Comment by loganmc10** [comment]
It could be some unimplemented GLES3 feature with the nouveau driver?

**Comment by loganmc10** [comment]
I can't seem to replay it on the Nvidia driver, it only replays full OpenGL traces

**Comment by orbea** [comment]
This is probably an issue in the nvc0 nouveau backend driver, it does not happen with llvmpipe or nv50 either. We should wait to see what the nouveau dev comes up with. :)

**Comment by orbea** [comment]
Here is some relevant information.
```
20:29 <imirkin> orbea: interesting. so actually it's the *road* that appears to be getting blended on there wrong. i think it's supposed to get culled out and it's not.
20:30 <orbea> Interesting
20:32 <imirkin> i think it has to do with polygon offsets
20:33 <imirkin> when i force-disable the polygon offset it doesn't look as bad
20:33 <imirkin> (but has lots of other nasty artifacts)
20:34 <orbea> is this entirely the nouveau backend or could it be the graphics plugin too (gliden64)?
20:35 <orbea> turning off the framebuffer emulation for gliden64 will fix it, but that is required to fix several other bugs
20:35 <imirkin> i mean ... gliden64 might be relying on something that's not well-specified
20:37 <imirkin> although given that it works on nv50 and llvmpipe and intel, i'm gonna guess it's something nouveau does poorly
```

**Comment by loganmc10** [comment]
There is a core option called "glupen64-PolygonOffsetFactor" that allows you to modify the polygon offset value, but it's only available for Android since that is how it was implemented in GLideN64

https://github.com/gonetz/GLideN64/pull/608

It was added because some Android devices have/had bugs related to Polygon offsets. You can google "glPolygonOffset bugs" to get an idea, maybe this is a similar problem?

**Comment by loganmc10** [comment]
You should see what the shadows in Mario 64 look like using GLES3, that should be a pretty good indicator if the polygon offset is the problem

**Comment by imirkin** [comment]
I think at least part of the reason is that the depth attachment is created as GL_DEPTH_COMPONENT32F - Z32F formats are much less predictable wrt polygon offsets.

**Comment by loganmc10** [comment]
I guess that would make sense too since this doesn't happen in GLES2 mode (which uses GL_DEPTH_COMPONENT), but the polygon offset stuff is the same in both modes.

**Comment by loganmc10** [comment]
@orbea You could try changing this line:

https://github.com/loganmc10/GLupeN64/blob/master/GLideN64/src/FBOTextureFormats.cpp#L46

to be ```GL_DEPTH_COMPONENT``` instead of ```GL_DEPTH_COMPONENT32F```. I'm not sure if that would even work but it would make it use the same setting as GLES2

**Comment by loganmc10** [comment]
if ```GL_DEPTH_COMPONENT``` doesn't work you could try ```GL_DEPTH_COMPONENT16```

**Comment by imirkin** [comment]
Looks like you use Z24 for desktop GL and gles2. Gles3 is a little picky
about using sized formats, but it's compatible with gles2 requirements iirc.

**Comment by orbea** [comment]
I see nothing wrong with Super Mario 64 and GLES3 including the previous context_reset shadows bug. I tried both of the proposed changes, `GL_DEPTH_COMPONENT` does not make a difference and `GL_DEPTH_COMPONENT16` will black screen with or without `depthFormatBytes = 2;`.

**Comment by imirkin** [comment]
You should make a new trace with GL_DEPTH_COMPONENT to help with debugging.

**Comment by loganmc10** [comment]
Even using Ubuntu inside VMware (so the VMware mesa driver) it looks fine, this seems to be a driver bug related to nouveau.

**Comment by imirkin** [comment]
The units of polygon offsets aren't extremely well-defined. Different HW may do different things with them, unfortunately. That said, I wouldn't exclude nvc0 setting things up slightly incorrectly. We do it the same way as on nv50 though, so it's odd that it would work on nv50 but not nvc0. Could be something else entirely.

Either way, a non-float-Z trace would be great.

**Comment by orbea** [comment]
@imirkin Here is a new trace with GL_DEPTH_COMPONENT.

http://ks392457.kimsufi.com/orbea/stuff/trace/libretro/Zelda%20OoT%2064_GL_DEPTH_COMPONENT_GLES3.trace.xz

**Comment by imirkin** [comment]
Grrr. The format selection rules are different for GLES.. for mesa:

   /* GLES allows the driver to choose any format which matches
    * the format+type combo, because GLES only supports unsized internal
    * formats and expects the driver to choose whatever suits it.
    */

@orbea - mind checking again with GL_DEPTH_COMPONENT24? (And make a new trace if it's still broken.)

**Comment by orbea** [comment]
Cool, that works! Doing that seems to entirely resolve the issue.
```
diff -ur GLupeN64/GLideN64/src/FBOTextureFormats.cpp GLupeN64/GLideN64/src/FBOTextureFormats.cpp
--- GLupeN64/GLideN64/src/FBOTextureFormats.cpp 2016-10-11 19:41:42.326528445 -0700
+++ GLupeN64/GLideN64/src/FBOTextureFormats.cpp 2016-10-11 19:37:51.585724981 -0700
@@ -43,9 +43,9 @@
        monochromeType = GL_UNSIGNED_BYTE;
        monochromeFormatBytes = 1;

-       depthInternalFormat = GL_DEPTH_COMPONENT32F;
+       depthInternalFormat = GL_DEPTH_COMPONENT24;
        depthFormat = GL_DEPTH_COMPONENT;
-       depthType = GL_FLOAT;
+       depthType = GL_UNSIGNED_INT;
        depthFormatBytes = 4;

        depthImageInternalFormat = GL_RGBA32F;
```

**Comment by loganmc10** [comment]
Fixed by 1b6240aac8f1ade071415c724522e8206b09b1be
